### PR TITLE
v2

### LINKED ISSUE #67: client_payload is too large
we did not change the workflow for a few days, but recently we see errors in the dispatching workflow:

image

the workflow file is:
```
name: Chatops

on:
  issue_comment:
    types: [created]

jobs:
  slash-command-dispatch:
    runs-on: ubuntu-latest
    steps:
      - name: Slash Command Dispatch
        uses: peter-evans/slash-command-dispatch@v1
        with:
          token: ${{ secrets.STAABM_TOKEN }}
          # reactions should be reported from github itself
          reaction-token: ${{ secrets.GITHUB_TOKEN }}
          commands: help, csfix, phpstan-baseline, churn
          issue-type: pull-request
```

is there anything we can do on our end, or can be done on slash-command-dispatch to fix this problem?
maybe its even a problem triggerd by the GithubAction's infrastructure/runtime?

### PR DESCRIPTION
**Pull request by peter-evans**
https://github.com/peter-evans/slash-command-dispatch/blob/dev/docs/updating.md#updating-from-v1-to-v2

Fixes #67, #68

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
 The action runs in `issue_comment` event workflows and checks the first line of comments for slash commands.
-When a valid command is found it creates a repository dispatch event that includes a payload containing full details of the command and its context.
+When a valid command is found it creates a [repository dispatch](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#repository_dispatch) event that includes a payload containing full details of the command and its context.
+It also supports creating [workflow dispatch](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch) events with defined input parameters.
 
-### Why repository dispatch?
+### Why create dispatch events?
 
 "ChatOps" with slash commands can work in a basic way by parsing the commands during `issue_comment` events and immediately processing the command.
 In repositories with a lot of activity, the workflow queue will get backed up very quickly trying to handle new `issue_comment` events *and* process the commands themselves.
```
Comment (clxmstaab) [inline README.md:16]: not sure you want to mention it, but another use case is:
- you can dispatch actions which are not directly related to the source code (those for which you don't need to checkout the repo) to a public repo in which running your actions does not count into your private github action minute budget while still beeing sure no private sources will be leaked

File: README.md [hunk 2]
```diff
@@ -90,13 +97,31 @@ You can use a [PAT](https://help.github.com/en/github/authenticating-to-github/c
...
+#### `dispatch-type`
+
+By default, the action creates [repository_dispatch](https://developer.github.com/v3/repos/#create-a-repository-dispatch-event) events.
+Setting `dispatch-type` to `workflow` will instead create [workflow_dispatch](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch) events.
+There are significant differences in the action's behaviour when using `workflow` dispatch. See [workflow dispatch](docs/workflow-dispatch.md) for usage details.
+
+For the majority of use cases, the default `repository` dispatch will likely be the most suitable for new workflows.
+If you already have `workflow_dispatch` workflows, you can execute them with slash commands using this action.
+
+| Repository Dispatch (default) | Workflow Dispatch |
```
Comment (clxmstaab) [inline README.md:119]: in the github docs, I can see the note

> Note: This event will only trigger a workflow run if the workflow file is on the master or default branch.

does this mean, when I initially setup slash commands for my repo..
- when using  `repository_dispatch` the slash commands only work after the PR has merged (so I can only test it after merging in the default-branch)
- but when using `workflow_dispatch` I can already test it in the PR without the need to merge it beforehand?

File: README.md [hunk 3]
```diff
@@ -90,13 +97,31 @@ You can use a [PAT](https://help.github.com/en/github/authenticating-to-github/c
...
+| --- | --- |
+| Events are created with a `client_payload` giving the target workflow access to a wealth of useful [context properties](https://github.com/peter-evans/slash-command-dispatch/tree/dev#accessing-contexts). | A `client_payload` cannot be sent with [workflow_dispatch](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch) events. The target workflow can only make use of up to 10 pre-defined inputs, the names of which must match named arguments supplied with the slash command. |
+| Slash commands can only execute workflows in the target repository's default branch. | Slash commands can execute workflows in any branch using the `ref` named argument. The reference can be a branch, tag, or a commit SHA. |
```
Comment (clxmstaab) [inline README.md:122]: > Slash commands can only execute workflows in the target repository's default branch.

this sounds like that this type would not be usefull in PRs, as I might assume while reading that I cannot run slash commands on my PRs branch..

File: docs/workflow-dispatch.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+To learn about the `workflow_dispatch` event, see GitHub's documentation on [manually running a workflow](https://docs.github.com/en/actions/configuring-and-managing-workflows/configuring-a-workflow#manually-running-a-workflow).
+
+## Action behaviour
+
+There are significant differences in the action's behaviour when using `workflow` dispatch.
+
+- When issuing a slash command with arguments, *only named arguments are accepted*. Unnamed arguments will be ignored.
+- A maximum of 10 named arguments are accepted. Additional named arguments after the first 10 will be ignored.
+- `ref` is a reserved named argument that does not count towards the maximum of 10. This is used to specify the target reference of the command. The reference can be a branch, tag, or a commit SHA. If you omit the `ref` argument, the target repository's default branch will be used.
+- A `client_payload` context cannot be sent with [workflow_dispatch](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#workflow_dispatch) events. The target workflow can only make use of up to 10 pre-defined inputs, the names of which must match named arguments supplied with the slash command.
```
Comment (clxmstaab) [inline docs/workflow-dispatch.md:14]: this paragraph perfectly describes the limitations of a `workflow_dispatch`.

what I am missing in the docs is a example/description why/when it makes sense to use `workflow_dispatch` over `repository_dispatch`. when does `workflow_dispatch` have a benefit over `repository_dispatch`?

File: README.md (hunk 1 above)
Comment (peter-evans) [reply README.md:16]: Added a note to mention this. Thanks.

File: README.md (hunk 2 above)
Comment (peter-evans) [reply README.md:121]: Yes, exactly. `workflow_dispatch` workflows will be executable in a PR branch before merge. You can use the `ref` named argument to target the PR branch name.

Comment (clxmstaab) [reply README.md:121]: I feel this is a major difference and should be noted

File: README.md [hunk 5]
```diff
@@ -17,6 +17,8 @@ In repositories with a lot of activity, the workflow queue will get backed up ve
 
 Dispatching commands to be processed elsewhere keeps the workflow queue moving quickly. It essentially enables parallel processing of workflows.
 
+A additional benefit of dispatching is that it allows non-sensitive workloads to be run in public repositories to save using private repository Action minutes.
```
Comment (clxmstaab) [inline README.md:20]:
```suggestion
A additional benefit of dispatching is that it allows non-sensitive workloads to be run in public repositories to save using private repository GitHub Action minutes.
```

File: docs/workflow-dispatch.md (hunk 4 above)
Comment (peter-evans) [reply docs/workflow-dispatch.md:14]: To be honest, I'm not sure myself yet. I wanted to add the feature and allow people to execute `workflow_dispatch` workflows with slash commands, but I'm not sure exactly why you would choose that workflow type over `repository_dispatch` if you were intending to execute it by slash command from the start. I think what is more likely is that users are going to start to create `workflow_dispatch` workflows to run manually from the UI. Then they might find this action and be interested in executing them by slash command. This feature allows them to do that without converting the workflow to `repository_dispatch`.

If additionally benefits become clear in the future, I'll update the docs. For now I hope the table at the following link will suffice.
https://github.com/peter-evans/slash-command-dispatch/tree/dev#dispatch-type

File: README.md (hunk 2 above)
Comment (peter-evans) [reply README.md:121]: It is a good benefit. Updated.

File: docs/workflow-dispatch.md (hunk 4 above)
Comment (clxmstaab) [reply docs/workflow-dispatch.md:14]: I see, thx for this insights.

**Pull request merged**